### ISSUE TITLE
`autograd_grad()` hangs when using custom autograd function.

### ISSUE THREAD

**Issue by rdinnager** [mlverse/torch#732]
I seem to have discovered a bug relating to custom autograd functions. If using a custom autograd function `autograd_grad()` hangs my R session indefinitely, seemingly doing nothing. Here is a reprex:

First I can confirm that `autograd_grad()` works with a normal `torch` function:

``` r
library(torch)
#> Warning: package 'torch' was built under R version 4.0.5

w <- torch_randn(4, 3, 5, 5)$requires_grad_()
w$retain_grad()
```

Using a regular torch function works as expected:

``` r
m <- torch_exp(w)$mean()
autograd_grad(m, w, torch_ones_like(w))
#> [[1]]
#> torch_tensor
#> (1,1,.,.) = 
#>    0.4243   0.6940   1.5162   1.9856   0.2466
#>    1.4057   0.8099   0.4362   0.3863   0.8204
#>    0.1427   1.4674   1.4824   2.0099   2.9860
#>    5.4964   1.8383   0.3746   0.6836   3.9882
#>    4.8878   1.8548  11.1562   0.4943   0.3239
#> 
#> (2,1,.,.) = 
#>   1.0517  0.5562  5.6161  0.5134  0.8257
#>   1.0828  0.6600  2.3329  1.2268  1.4636
#>   1.8087  1.2369  0.2794  1.2175  3.0913
#>   1.1161  1.2489  0.1120  0.5173  4.6910
#>   1.5832  0.0965  0.1686  1.9422  0.6292
#> 
#> (3,1,.,.) = 
#>   1.6194  0.1546  0.9133  0.1238  0.7758
#>   0.3842  2.5723  4.0663  1.0213  1.1482
#>   2.7690  0.8975  6.5983  0.2942  2.9477
#>   1.6238  0.0875  0.2772  0.1561  0.7254
#>   0.3599  0.8841  3.1817  4.8949  1.7158
#> 
#> (4,1,.,.) = 
#>   4.2203  4.3423  3.0568  0.5859  1.6503
#>   2.0431  1.6915  1.6390  0.2918  0.8289
#>   0.5754  0.5292  0.4267  1.2777  0.3169
#>   1.6316  0.8022  1.3314  1.8562  1.6474
#>   7.8410  0.0994  2.7157  2.0904  0.4756
#> 
#> (1,2,.,.) = 
#>   0.7499  1.1368  3.3558  0.6498  0.4081
#> ... [the output was truncated (use n=-1 to disable)]
#> [ CPUFloatType{4,3,5,5} ]
```

Now we create a custom autograd function (this is from the example for `autograd_function`), with print statements so we know when the forward and backward are being called.

``` r
exp2 <- autograd_function(
  forward = function(ctx, i) {
    print("exp2 forward..")
    result <- i$exp()
    ctx$save_for_backward(result = result)
    result
  },
  backward = function(ctx, grad_output) {
    print("exp2 backward..")
    list(i = grad_output * ctx$saved_variables$result)
  }
)
```

Now we try using it (using a new tensor just to make sure there are no leftover gradients from before).

``` r
w2 <- torch_randn(4, 3, 5, 5)$requires_grad_()
w2$retain_grad()
m2 <- exp2(w2)$mean()
#> [1] "exp2 forward.."
```

Now if we just try to run `autograd_backward` it seems to work fine, which is good.

``` r
w2$grad
#> torch_tensor
#> [ Tensor (undefined) ]
autograd_backward(m2)
#> [1] "exp2 backward.."
w2$grad
#> torch_tensor
#> (1,1,.,.) = 
#>  0.01 *
#>   1.4296  0.2108  0.0845  0.1505  0.3136
#>    0.1364  0.2412  0.2017  0.1687  0.1807
#>    0.4484  0.2436  0.1247  1.1385  0.0985
#>    0.5036  0.0624  0.2946  3.1731  0.6869
#>    0.7946  0.1512  0.3405  2.3405  0.4789
#> 
#> (2,1,.,.) = 
#>  0.01 *
#>   0.1612  0.2102  0.0832  0.1444  1.4350
#>    0.2517  0.1590  2.3417  0.7447  0.0924
#>    0.0861  0.1698  0.4182  0.0933  0.1268
#>    0.1288  0.2460  0.6827  0.1927  0.5831
#>    0.0824  0.4004  0.1739  0.9243  0.2354
#> 
#> (3,1,.,.) = 
#>  0.01 *
#>   0.6912  1.3207  0.2361  0.2429  0.9487
#>    2.3959  0.3300  1.7289  0.1317  0.0583
#>    0.9818  0.2520  0.1574  0.4889  1.3069
#>    0.1529  0.8999  0.2195  0.1458  0.1667
#>    0.1352  0.1531  0.2509  0.3607  0.5693
#> 
#> (4,1,.,.) = 
#>  0.01 *
#>   0.1514  0.1023  0.3624  1.9982  1.9803
#>    0.6523  0.3665  4.0502  0.4030  1.1081
#>    1.7496  2.8275  0.1995  0.2265  0.6489
#>    0.5347  0.6367  0.2195  0.9767  0.5406
#> ... [the output was truncated (use n=-1 to disable)]
#> [ CPUFloatType{4,3,5,5} ]
```

But if we try and calculate a gradient using `autograd_grad` instead, the function hangs forever, with no error message(s). The print statement I placed in forward and backward never get printed, implying they simply are not being run.

``` r
w3 <- torch_randn(4, 3, 5, 5)$requires_grad_()
w3$retain_grad()
m3 <- exp2(w3)$mean()
#> [1] "exp2 forward.."
```

If I run this then my session hangs:

``` r
autograd_grad(m3, w3, torch_ones_like(w3))
```

Note that this happens also if I do this all on a GPU, so it is not specific to the CPU calculations.
My session info:

``` r
sessionInfo()
#> R version 4.0.2 (2020-06-22)
#> Platform: x86_64-w64-mingw32/x64 (64-bit)
#> Running under: Windows 10 x64 (build 22000)
#> 
#> Matrix products: default
#> 
#> locale:
#> [1] LC_COLLATE=English_Australia.1252  LC_CTYPE=English_Australia.1252   
#> [3] LC_MONETARY=English_Australia.1252 LC_NUMERIC=C                      
#> [5] LC_TIME=English_Australia.1252    
#> 
#> attached base packages:
#> [1] stats     graphics  grDevices utils     datasets  methods   base     
#> 
#> other attached packages:
#> [1] torch_0.6.0
#> 
#> loaded via a namespace (and not attached):
#>  [1] Rcpp_1.0.7        knitr_1.36        magrittr_2.0.1    bit_4.0.4        
#>  [5] R6_2.5.1          rlang_0.4.11      fansi_0.5.0       stringr_1.4.0    
#>  [9] styler_1.5.1      highr_0.9         tools_4.0.2       xfun_0.25        
#> [13] coro_1.0.1        utf8_1.2.2        withr_2.4.2       htmltools_0.5.1.1
#> [17] ellipsis_0.3.2    bit64_4.0.5       yaml_2.2.1        digest_0.6.27    
#> [21] tibble_3.1.3      lifecycle_1.0.0   crayon_1.4.2      processx_3.5.2   
#> [25] callr_3.7.0       purrr_0.3.4       ps_1.6.0          vctrs_0.3.8      
#> [29] fs_1.5.0          glue_1.4.2        evaluate_0.14     rmarkdown_2.10   
#> [33] reprex_2.0.1      stringi_1.7.3     compiler_4.0.2    pillar_1.6.2     
#> [37] backports_1.2.1   pkgconfig_2.0.3
```

<sup>Created on 2021-11-05 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by dfalbel** [comment]
Indeed! I think we are deadlocking waiting for some code in autograd to run. This is one of the trickiest parts of the interaction between the R code and LibTorch. I will take a look before the next release.

**Comment by rdinnager** [comment]
Just checked this again with the dev version after recent autograd updates to fix memory leak, and can confirm that this issue still happens. Not surprising, since you weren't trying to fix this, but just letting you know it didn't get incidentally resolved.

**Comment by dfalbel** [comment]
Yeah, I found out what the problem is though. The issue is that we didn't implement the same strategy as for `backward` for autograd:

https://github.com/mlverse/torch/blob/b6339f71f2b359e56c92777fa93e8ac7774361ae/src/autograd.cpp#L466-L481

We have to do something similar to what we do with backward. ie launching the grad call in a separate thread and leaving the main thread waiting to execute calls to the main thread that might be scheduled when running grad. 

See:

https://github.com/mlverse/torch/blob/b6339f71f2b359e56c92777fa93e8ac7774361ae/src/autograd.cpp#L138-L188

I will work on this soon, but waiting to see if #754 can simplify that code path before.

**Comment by rdinnager** [comment]
Ah, I see. Great news that you found the problem! Autograd functional relies on `autograd_grad()` for much of its functionality so it would be great to get this working, so that custom autograd functions won't cause a problem for it. But there is no rush, the autograd functional implementation will probably take me awhile to get through at my current rate.
